lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,379 @@
+
+    INFO("Resetting ring buffer memory at position time_ind: {}\n", time_ind);
+    // set the full time_ind to 0 in the ring buffer
+    std::vector<char> zeros(nfreq * frame_size, 0);
+    char* tmp = zeros.data();
+    memcpy(buf_write_pos, &tmp, sizeof(tmp));
+
+    INFO("Memory reset\n");
+}
+
+void VisSharedMemWriter::write_to_memory(const visFrameView& frame, uint32_t time_ind, uint32_t freq_ind) {
+    // write frame to ring buffer at time_ind and freq_ind
+
+    uint8_t *buf_write_pos = buf_addr + ((time_ind * nfreq + freq_ind) * frame_size);
+    int64_t *access_record_write_pos = access_record_addr + (time_ind * nfreq + freq_ind);
+
+    DEBUG("Writing ringbuffer to time_ind {} and freq_ind {}\n", time_ind, freq_ind);
+
+    // notify that time_ind and freq_ind are being written to, by setting that
+    // location to invalid in the access record
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}. Dropping frame.\n", _sem_name);
+        return;
+    }
+    memcpy(access_record_write_pos, &invalid, sizeof(invalid));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+    // first write the metadata, then the data, then the valid byte
+    // add valid_size amount of padding
+    memcpy(buf_write_pos + valid_size, frame.metadata(), metadata_size);
+    memcpy(buf_write_pos + metadata_size + valid_size, frame.data(), data_size);
+    memcpy(buf_write_pos, &valid, sizeof(valid));
+
+    // Document the fpga sequence counter for that frame in the access record
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
+
+
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {} before metadata was updated.\n", _sem_name);
+        return;
+    }
+
+    DEBUG("Writing fpga_seq {} to time index {}\n", fpga_seq, time_ind);
+    memcpy(access_record_write_pos, &fpga_seq, sizeof(fpga_seq));
+
+    // update num_writes
+    num_writes++;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+}
+
+void VisSharedMemWriter::main_thread() {
+    DEBUG("Reached main thread\n");
+
+    frameID frame_id(in_buf);
+
+    // The current position in the ring buffer of the most recent time sample
+    // from 0 -> _ntime
+    cur_pos = modulo<int>(_ntime);
+
+    // Create the semaphore, and gain first access to it
+    sem = sem_open(
+            _sem_name.c_str(),
+            (O_CREAT | O_EXCL),
+            (S_IRUSR | S_IWUSR),
+            1
+        );
+
+    if (sem == SEM_FAILED) {
+        FATAL_ERROR("Failed to create semaphore {}", _sem_name);
+        return;
+    }
+
+    DEBUG("Semaphore created.\n");
+
+    // Acquire semaphore until shared memory is created
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}", _sem_name);
```

Comment:
Anywhere you would print an `ERROR` and return should probably be a `FATAL_ERROR`, as you've already decided you can't recover from it.